genienlp/models/base.py

```diff
@@ -77,3 +79,8 @@ def add_new_vocab_from_data(self, tasks, resize_decoder=False):
         self.numericalizer.grow_vocab(tasks)
         if self.numericalizer.num_tokens > old_num_tokens:
             logger.info(f'Vocabulary has expanded to {self.numericalizer.num_tokens} tokens')
+
+    def set_task_dependent_generation_kwargs(self, tasks):
+        self._output_scores = any('loss' in task.metrics for task in tasks)
+        self._output_attentions = any(isinstance(task, Translate) for task in tasks)
+        self._output_hidden_states = False
```

Comment:
If this is task dependent, then surely it should live on the Task, and not on the model?
More generally, model code should be independent on the task.

Also, should we just unconditionally set these to true (except _output_scores maybe)? Is there a performance loss if we do that?

Reply:
> If this is task dependent, then surely it should live on the Task, and not on the model?

We're already passing tasks to model init for numericalizer to read the special tokens. Also not sure why it should be in Task. Based on task, we need to adjust some attributes of the model. We still need to pass some information from task to model if not the whole object.

> Also, should we just unconditionally set these to true (except _output_scores maybe)? Is there a performance loss if we do that?

It makes a difference in memory usage. With all of them being False, we can fit bigger batches in one pass.

Reply:
You don't want to depend on the specific Task instance, because next time someone will come along with another Translate-like class and it won't work for obscure reasons.
If you want, make it an attribute of the base Task class like `needs_attention_scores` or similar.